library/ssl_tls13_server.c

```diff
@@ -24,16 +24,770 @@
+        curve_info = mbedtls_ecp_curve_info_from_tls_id( tls_grp_id );
+
+        if( curve_info != NULL )
+        {
+
+            MBEDTLS_SSL_DEBUG_MSG( 4, ( "supported curve: %s", curve_info->name ) );
+            /*
+             * Here we only update offered_group_id field with the first
+             * offered group
+             */
+            ssl->handshake->offered_group_id = tls_grp_id;
+            break;
+        }
+
+        p += 2;
+    }
+
+    return( 0 );
+
+}
+#endif /* MBEDTLS_ECDH_C */
+
+#if defined(MBEDTLS_ECDH_C)
+/*
+ *  ssl_tls13_parse_key_shares_ext() verifies whether the information in the
+ *  extension is correct and stores the provided key shares. Whether this is an
+ *  acceptable key share depends on the selected ciphersuite.
+ *
+ *  Possible return values are:
+ *  - 0: Successful processing of the client provided key share extension.
+ *  - MBEDTLS_ERR_SSL_HRR_REQUIRED: The key share provided by the client
+ *    does not match a group supported by the server. A HelloRetryRequest will
+ *    be needed.
+ *  - Another negative return value for fatal errors.
+*/
+
+static int ssl_tls13_parse_key_shares_ext( mbedtls_ssl_context *ssl,
+                                           const unsigned char *buf,
+                                           const unsigned char *end )
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    unsigned char const *p = buf;
+    unsigned char const *client_shares_end;
+    size_t client_shares_len, key_exchange_len;
+    int match_found = 0;
+
+    /* From RFC 8446:
+     *
+     * struct {
+     *     KeyShareEntry client_shares<0..2^16-1>;
+     * } KeyShareClientHello;
+     *
+     */
+
+    MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, 2 );
+    client_shares_len = MBEDTLS_GET_UINT16_BE( p, 0 );
+    p += 2;
+    MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, client_shares_len );
+
+    ssl->handshake->offered_group_id = 0;
+    client_shares_end = p + client_shares_len;
+
+    /* We try to find a suitable key share entry and copy it to the
+     * handshake context. Later, we have to find out whether we can do
+     * something with the provided key share or whether we have to
+     * dismiss it and send a HelloRetryRequest message.
+     */
+
+    for( ; p < client_shares_end; p += key_exchange_len )
+    {
+        uint16_t group;
+
+        /*
+         * struct {
+         *    NamedGroup group;
+         *    opaque key_exchange<1..2^16-1>;
+         * } KeyShareEntry;
+         */
+        MBEDTLS_SSL_CHK_BUF_READ_PTR( p, client_shares_end, 4 );
+        group = MBEDTLS_GET_UINT16_BE( p, 0 );
+        p += 2;
+        key_exchange_len = MBEDTLS_GET_UINT16_BE( p, 0 );
+        p += 2;
+        MBEDTLS_SSL_CHK_BUF_READ_PTR( p, client_shares_end, key_exchange_len );
+
+        /* Continue parsing even if we have already found a match,
+         * for input validation purposes.
+         */
+        if( match_found == 1 )
+            continue;
+
```

Comment:
Here we need also (as in the parsing of the supported groups extension) to check that the group is supported and offered, see https://github.com/Mbed-TLS/mbedtls/pull/5707/commits/2c6076418223bc1fc6ce3002ca8d696752ba774a.

Reply:
Not addressed

Reply:
Here you need something like:
```
if( ! mbedtls_ssl_named_group_is_offered( ssl, group ) ||         
    ! mbedtls_ssl_named_group_is_supported( group ) )                               
{                                                                       
    continue;                                                           
}    
```
The "is_supported" check is then done independently of the type of the group and we can probably remove the same type of check in the ECDHE code below. Thus all together it would be something like:
```
if( ! mbedtls_ssl_named_group_is_offered( ssl, group ) ||         
    ! mbedtls_ssl_named_group_is_supported( group ) )                               
{                                                                       
    continue;                                                           
}
curve_info = mbedtls_ecp_curve_info_from_tls_id( group );

if( mbedtls_ssl_tls13_named_group_is_ecdhe( group ) )                   
{                                                                       
    match_found = 1;                                                    
    MBEDTLS_SSL_DEBUG_MSG( 2, ( "ECDH curve: %s", curve_info->name ) ); 
    ret = mbedtls_ssl_tls13_read_public_ecdhe_share(                    
                    ssl, p - 2, key_exchange_len + 2 );                         
    if( ret != 0 )                                                      
        return( ret );                                                  
}                                                                        
```

Reply:
"Here we need also (as in the parsing of the supported groups extension) to check that the group is supported and offered, see https://github.com/Mbed-TLS/mbedtls/commit/2c6076418223bc1fc6ce3002ca8d696752ba774a."
Not addressed.